### PR TITLE
20241217 3187 add another device qr

### PR DESCRIPTION
**Pull request by Firstyear**
Trying something different this time crabs,

This is two commits - one to improve cookie handling which I realised had some issues in the process of working on re-enabling the enrollment flow. The other part is re-adding the "scan a qr code" to enroll another device flow. 

Fixes #3187 

Checklist

- [x] This PR contains no AI generated code
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

### CODE REVIEW

**Comment by Firstyear** [comment]
Pre-apologies to @CEbbinghaus that it's not 1 line.

File: server/core/src/https/views/constants.rs [hunk 1]
```diff
@@ -5,6 +5,7 @@ use serde::{Deserialize, Serialize};
 pub(crate) enum ProfileMenuItems {
     UserProfile,
     Credentials,
+    EnrollDevice,
```
Comment (yaleman) [inline server/core/src/https/views/constants.rs:8]: This is Authstralia

```suggestion
    EnrolDevice,
```

File: server/core/src/https/views/constants.rs [hunk 2]
```diff
@@ -25,6 +26,7 @@ pub(crate) enum Urls {
     Apps,
     CredReset,
     CredResetError,
+    EnrollDevice,
```
Comment (yaleman) [inline server/core/src/https/views/constants.rs:29]: This is Authstralia

```suggestion
    EnrolDevice,
```

File: server/core/src/https/views/constants.rs [hunk 3]
```diff
@@ -38,6 +40,7 @@ impl AsRef<str> for Urls {
             Self::Apps => "/ui/apps",
             Self::CredReset => "/ui/reset",
             Self::CredResetError => "/ui/reset/err",
+            Self::EnrollDevice => "/ui/enroll",
```
Comment (yaleman) [inline server/core/src/https/views/constants.rs:43]: This is Authstralia

```suggestion
            Self::EnrolDevice => "/ui/enrol",
```

File: server/core/src/https/views/enroll.rs [hunk 4]
```diff
@@ -0,0 +1,115 @@
+use askama::Template;
```
Comment (yaleman) [inline server/core/src/https/views/enroll.rs:1]: This is Authstralia, please rename the file to `enrol.rs`

File: server/core/templates/user_settings_partial_base.html [hunk 5]
```diff
@@ -20,6 +20,8 @@
             ProfileMenuItems::UserProfile, "person") %)
             (% call side_menu_item("Credentials", (Urls::UpdateCredentials),
             ProfileMenuItems::Credentials, "shield-lock") %)
+            (% call side_menu_item("Enroll Device", (Urls::EnrollDevice),
```
Comment (yaleman) [inline server/core/templates/user_settings_partial_base.html:23]: This is Authstralia

```suggestion
            (% call side_menu_item("Enrol Device", (Urls::EnrolDevice),
```

File: server/core/templates/user_settings_partial_base.html [hunk 6]
```diff
@@ -20,6 +20,8 @@
...
+            ProfileMenuItems::EnrollDevice, "phone-flip") %)
```
Comment (yaleman) [inline server/core/templates/user_settings_partial_base.html:24]: This is Authstralia

```suggestion
            ProfileMenuItems::EnrolDevice, "phone-flip") %)
```

File: server/core/templates/enroll_device.html [hunk 7]
```diff
@@ -0,0 +1,20 @@
+(% extends "user_settings_partial_base.html" %)
+
+(% block selected_setting_group %)
+Enroll Another Device
+(% endblock %)
+
+(% block settings_window %)
+<p>You can enroll another device to your account by scanning the QR code or following the link below.</p>
```
Comment (yaleman) [inline server/core/templates/enroll_device.html:8]: This is Authstralia

```suggestion
<p>You can enrol another device to your account by scanning the QR code or following the link below.</p>
```

File: server/core/templates/enroll_device.html [hunk 8]
```diff
@@ -0,0 +1,20 @@
+(% extends "user_settings_partial_base.html" %)
+
+(% block selected_setting_group %)
+Enroll Another Device
```
Comment (yaleman) [inline server/core/templates/enroll_device.html:4]: This is Authstralia

```suggestion
Enrol Another Device
```

File: server/core/src/https/views/mod.rs [hunk 9]
```diff
@@ -37,6 +38,7 @@ pub fn view_router() -> Router<ServerState> {
             get(|| async { Redirect::permanent(Urls::Login.as_ref()) }),
         )
         .route("/apps", get(apps::view_apps_get))
+        .route("/enroll", get(enroll::view_enroll_get))
```
Comment (yaleman) [inline server/core/src/https/views/mod.rs:41]: This is Authstralia

```suggestion
        .route("/enrol", get(enroll::view_enroll_get))
```

File: server/core/src/https/views/mod.rs [hunk 10]
```diff
@@ -16,6 +16,7 @@ use crate::https::ServerState;
 mod apps;
 mod constants;
 mod cookies;
+mod enroll;
```
Comment (yaleman) [inline server/core/src/https/views/mod.rs:19]: This is Authstralia

```suggestion
mod enrol;
```

File: server/core/src/https/views/enroll.rs [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+#[derive(Template)]
+#[template(path = "enroll_device.html")]
+pub(crate) struct EnrollDeviceView {
+    menu_active_item: ProfileMenuItems,
+    secret: String,
+    qr_code_svg: String,
+    uri: Url,
+}
+
+pub(crate) async fn view_enroll_get(
```
Comment (yaleman) [inline server/core/src/https/views/enroll.rs:42]: This is Authstralia

```suggestion
pub(crate) async fn view_enrol_get(
```

File: server/core/src/https/views/enroll.rs [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+#[derive(Template)]
+#[template(path = "user_settings.html")]
+struct ProfileView {
+    navbar_ctx: NavbarCtx,
+    profile_partial: EnrollDeviceView,
+}
+
+#[derive(Template)]
+#[template(path = "enroll_device.html")]
```
Comment (yaleman) [inline server/core/src/https/views/enroll.rs:34]:
```suggestion
#[template(path = "enrol_device.html")]
```

File: server/core/src/https/views/enroll.rs [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+pub(crate) struct EnrollDeviceView {
```
Comment (yaleman) [inline server/core/src/https/views/enroll.rs:35]: This is Authstralia

```suggestion
pub(crate) struct EnrolDeviceView {
```

**Review by yaleman — DISMISSED** [review]
I think you get the idea 🇦🇺 🦀 🤔

**Comment by CEbbinghaus** [comment]
@yaleman's review style
giphy

File: server/core/src/https/views/enroll.rs [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+    let uat: UserAuthToken = state
+        .qe_r_ref
+        .handle_whoami_uat(client_auth_info.clone(), kopid.eventid)
+        .await
+        .map_err(|op_err| HtmxError::new(&kopid, op_err))?;
+
+    let time = time::OffsetDateTime::now_utc() + time::Duration::new(60, 0);
+    let can_rw = uat.purpose_readwrite_active(time);
+
+    if can_rw {
```
Comment (CEbbinghaus) [inline server/core/src/https/views/enroll.rs:58]: Would be nice to `!can_rw` and remove the else. Save some whitespace with the bigger block

File: server/core/templates/credentials_update_partial.html [hunk 15]
```diff
@@ -147,7 +147,8 @@ <h4>UNIX Password</h4>
                     <div class="mt-2 pt-2 border-top">
                         <button class="btn btn-danger"
                             hx-post="/ui/api/cu_cancel"
-                            hx-target="#main">Discard Changes</button>
+                            hx-boost="false"
+                        >Discard Changes</button>
```
Comment (CEbbinghaus) [inline server/core/templates/credentials_update_partial.html:151]:
```suggestion
                            hx-boost="false">
                                Discard Changes
                            </button>
```
Doesn't look like we have a consistent formatting style for HTML. Might be good to standardize so it's not whatever the dev (or reviewer) feels like

File: server/core/templates/enroll_device.html [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+(% block settings_window %)
+<p>You can enroll another device to your account by scanning the QR code or following the link below.</p>
+
+<div id="intentInfo">
+    <div>((qr_code_svg|safe))</div>
+    <ul>
+        <li>Url: <code>((uri|safe))</code></li>
+        <li>Secret: <code>(( secret ))</code></li>
+    </ul>
+
```
Comment (CEbbinghaus) [inline server/core/templates/enroll_device.html:16]:
```suggestion
```

File: server/core/templates/enroll_device.html [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+</div>
+
```
Comment (CEbbinghaus) [inline server/core/templates/enroll_device.html:18]:
```suggestion
```

**Review by CEbbinghaus — CHANGES_REQUESTED** [review]
Just a few stylistic things. Feel free to ignore

File: server/core/templates/credentials_update_partial.html (hunk 15 above)
Comment (Firstyear) [reply server/core/templates/credentials_update_partial.html:151]: Yep, I think we should do this is a separate issue to get that sorted. I know I'm a culprit here too.

File: server/core/src/https/views/enrol.rs [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+    let uat: UserAuthToken = state
+        .qe_r_ref
+        .handle_whoami_uat(client_auth_info.clone(), kopid.eventid)
+        .await
+        .map_err(|op_err| HtmxError::new(&kopid, op_err))?;
+
+    let time = time::OffsetDateTime::now_utc() + time::Duration::new(60, 0);
+    let can_rw = uat.purpose_readwrite_active(time);
+
+    // The user lacks an elevated session, request a re-auth.
```
Comment (CEbbinghaus) [inline server/core/src/https/views/enrol.rs:58]: Love the context, This is the kind of comment I am here for <3

**Review by CEbbinghaus — APPROVED** [review]
This is Authstralia, (and a bloody good PR too 😉)

**Pull request merged**